operator/internal/manifests/storage/schema_test.go

```diff
@@ -0,0 +1,145 @@
+package storage
+
+import (
+	"testing"
+	"time"
+
+	lokiv1beta1 "github.com/grafana/loki/operator/api/v1beta1"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpdateSchema_AddNewSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    time.Now().Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV12,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV12, []Schema{})
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_AppendNewSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+		{
+			From:    time.Now().AddDate(0, 0, DaysToUpdate).Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV12,
+		},
+	}
+
+	test := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV12, test)
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_IgnoreSameSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	test := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV11, test)
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_ModifyUnappliedSchema(t *testing.T) {
```

Comment:
This test case assumes in the `test` schema that we have a partially applies schema for `v10`. This case should be by all means not possible. The reconciliation loop should either fail or succeed on each individual create/update op to the apiserver but never store partial results. Thus what is the purpose of this test?

Reply:
This test was to make sure that the operator changes the last element if for some reason a user flips between v11 and v12 on the same day. However when doing examples by hand, I found it difficult to write this case without a third version. I agree this is not a realistic test; it's purely to test the functional case.